core/shared/src/main/scala/zio/ZManaged.scala

```diff
@@ -1064,27 +1064,40 @@ final class ZManaged[-R, +E, +A] private (reservation: ZIO[R, E, Reservation[R,
    */
   def zipWithPar[R1 <: R, E1 >: E, A1, A2](that: ZManaged[R1, E1, A1])(f0: (A, A1) => A2): ZManaged[R1, E1, A2] =
     ZManaged[R1, E1, A2] {
-      Ref.make[List[Exit[Any, Any] => ZIO[R1, Nothing, Any]]](Nil).map { finalizers =>
+      Ref.make[Either[Exit[Any, Any], List[Exit[Any, Any] => ZIO[R1, Nothing, Any]]]](Right(Nil)).map { ref =>
         Reservation(
           acquire = {
             val left = ZIO.uninterruptibleMask { restore =>
-              reserve
-                .flatMap(res => finalizers.update(fs => res.release :: fs).as(res))
-                .flatMap(res => restore(res.acquire))
+              self.reserve.flatMap { reservation =>
+                ref.updateSomeAndGet {
+                  case Right(finalizers) => Right(reservation.release :: finalizers)
+                }.flatMap {
+                  case Left(exit) => restore(reservation.acquire).ensuring(reservation.release(exit))
```

Comment:
Very tricky indeed. Nice catch. Happy to see it's a relatively esoteric scenario.